app/react/Forms/components/MultiSelect.tsx

```diff
@@ -83,25 +82,70 @@ abstract class MultiSelectBase<ValueType> extends Component<
     return [];
   }
 
-  private searchInputRef: RefObject<HTMLInputElement>;
+  abstract getCheckedList(): string[];
+
+  abstract markUnchecked(value: ValueType, option: Option): ValueType;
+
+  abstract markChecked(value: ValueType, option: Option): ValueType;
+
+  private getGroupStateValue(value: ValueType, group: Option) {
+    let newValue = this.markChecked(value, group);
+    group.options!.forEach(_item => {
+      if (this.checked(_item) !== SelectStates.OFF) {
+        newValue = this.markUnchecked(newValue, _item);
+      }
+    });
+
+    return newValue;
+  }
+
+  private getOnStateValue(value: ValueType, group: Option) {
+    let newValue = this.markUnchecked(value, group);
+    group.options!.forEach(_item => {
+      if (this.checked(_item) !== SelectStates.ON) {
+        newValue = this.markChecked(newValue, _item);
+      }
+    });
+
+    return newValue;
+  }
+
+  private getOffStateValue(value: ValueType, group: Option) {
+    let newValue = this.markUnchecked(value, group);
+    group.options!.forEach(_item => {
+      if (this.checked(_item) !== SelectStates.OFF) {
+        newValue = this.markUnchecked(newValue, _item);
+      }
+    });
+
+    return newValue;
+  }
 
   changeGroup(group: Option, e: React.ChangeEvent<HTMLInputElement>) {
     let { value } = this.props;
-    if (e.target.checked) {
-      group.options!.forEach(_item => {
-        if (this.checked(_item) !== SelectStates.ON) {
-          value = this.markChecked(value, _item);
-        }
-      });
+    const { allowSelectGroup } = this.props;
+    const previousState = parseInt(e.target.dataset.state!, 10);
+
+    if (previousState === SelectStates.OFF) {
```

Comment:
Maybe there is a way to avoid that many 'ifs'. Seems that needs to be extracted to a new class and you can use a command pattern or something like this:

  changeGroup(group: Option, e: React.ChangeEvent<HTMLInputElement>) {
    const { value } = this.props;
    const { allowSelectGroup } = this.props;
    const previousState = parseInt(e.target.dataset.state!, 10);

    if (allowSelectGroup && previousState === SelectStates.ON) {
      this.props.onChange(this.getGroupStateValue(value, group));
      return;
    }

    const getNewValueFromPreviousState: { [key: string]: any } = {
      [SelectStates.ON]: this.getOffStateValue(value, group),
      [SelectStates.OFF]: this.getOnStateValue(value, group),
      [SelectStates.GROUP]: this.getOffStateValue(value, group),
      [SelectStates.PARTIAL]: this.getOffStateValue(value, group),
    };

    this.props.onChange(getNewValueFromPreviousState[previousState]);
  }

Reply:
Yeah, which is the equivalent of using a `switch` statement. I thought about revisiting this, and forgot it :D let me give it a try. Thanks for noticing this.

Reply:
Great :)